jax_md/elasticity.py

```diff
@@ -0,0 +1,582 @@
+  if not (DIM==2 or DIM==3):
+    raise AssertionError('DIM must be 2 or 3')
+  
+  rank = len(T.shape)
+  if not (rank==2 or rank==4):
+    raise AssertionError('T must have rank 2 or 4')
+  
+  m_map, weight = _get_mandel_mapping_weight(DIM)
+  
+  if rank == 2:
+    extract = lambda idx, w: T[idx[0], idx[1]] * w
+    M = vmap(extract, in_axes=(0,0))(m_map, weight)
+  else:
+    extract = lambda idx0, idx1, w0, w1: T[idx0[0], idx0[1], idx1[0], idx1[1]] * w0 * w1
+    M = vmap(vmap(extract, in_axes=(0,None,0,None)), in_axes=(None,0,None,0))(m_map, m_map, weight, weight)
+  return M
+
+def mandel_to_tensor(M: Array) -> Array:
+  """ Perform the inverse of M = tensor_to_mandel(T).
+
+  Args:
+    M: Array of shape (3,), (6,), (3,3), or (6,6)
+  
+  Output: Array of shape (2,2), (3,3), (2,2,2,2), or (3,3,3,3)
+  """
+  DIM = M.shape[0]
+  if not (DIM==3 or DIM==6):
+    raise AssertionError('DIM must be 3 or 6')
+  
+  rank = len(M.shape)
+  if not (rank==1 or rank==2):
+    raise AssertionError('T must have rank 1 or 2')
+
+  def mandel_index(i,j):
+    return lax.cond(i==j, lambda ij: ij[0], lambda ij: DIM-ij[0]-ij[1], (i,j))
+
+  if DIM == 3:
+    dimension = 2
+  else:
+    dimension = 3
+
+  tensor_range = jnp.arange(dimension)
+  _, weight = _get_mandel_mapping_weight(dimension)
+
+  if rank == 1:
+    def extract(i,j):
+      idx = mandel_index(i,j)
+      return M[idx] / weight[idx]
+    T = vmap(vmap(extract, in_axes=(0,None)), in_axes=(None,0))(tensor_range, tensor_range)
+  else:
+    def extract(i,j,k,l):
+      idx0 = mandel_index(i,j)
+      idx1 = mandel_index(k,l)
+      return M[idx0,idx1] / (weight[idx0] * weight[idx1])
+    T = vmap(vmap(vmap(vmap(extract, in_axes=(0,None,None,None)), in_axes=(None,0,None,None)), in_axes=(None,None,0,None)), in_axes=(None,None,None,0))(tensor_range, tensor_range, tensor_range, tensor_range)
+  return T
+
+
+
+@partial(jit,static_argnums=(1,))
+def _extract_elements(C, as_dict):
+  """ Convert an elastic modulus tensor into a list of 6 (21) unique elements
+        in 2 (3) dimensions
+      
+      in 2d, these are:
+      cxxxx,cyyyy,cxyxy,cxxyy,cxxxy,cyyxy
+
+      in 3d, these are:
+      cxxxx,cyyyy,czzzz,cyzyz,cxzxz,cxyxy,cyyzz,cxxzz,cxxyy,cxxyz,cxxxz,cxxxy,cyyyz,cyyxz,cyyxy,czzyz,czzxz,czzxy,cyzxz,cyzxy,cxzxy
+
+  Args:
+    C: A previously calculated elastic modulus tensor represented as an 
+      array of shape (spatial_dimension,spatial_dimension,spatial_dimension,
+      spatial_dimension), where spatial_dimension is either 2 or 3. C must 
+      satisfy both the major and minor symmetries, but this is not checked.
+    as_dict: boolean. If true, return a dictionary with interpretable keys.
+      If false, return an array that follows an internal convention.
+  """
+  if C.shape[0] == 2:
+    indices = ( [0, 1, 0, 0, 0, 0],
+                [0, 1, 1, 0, 0, 1],
+                [0, 1, 0, 1, 0, 1],
+                [0, 1, 1, 1, 1, 1])
+    clist = C[ indices ] 
+    if as_dict==True:
```

Comment:
What are the conventions for equality checking? flake8 tells me:
"flake8(E712): comparison to True should be 'if cond is True:' or 'if cond:' "